File: day9-visualization/scripts/R/heatmap.R
# ------------------------------------------------------------
# TOY MAG ABUNDANCE HEATMAP (clustered)
#    Rows = MAGs, Cols = Samples, Values = abundance
# ------------------------------------------------------------
library(tidyverse)
library(pheatmap)
library(vegan)
library(GGally)
library(plotly)


n_mags <- 30
mag_ids <- sprintf("MAG_%02d", 1:n_mags)

samples2 <- c(
  "Spring_Low_FL","Spring_Med_FL","Spring_High_FL",
  "Spring_Low_PA","Spring_Med_PA","Spring_High_PA",
  "Summer_Low_FL","Summer_Med_FL","Summer_High_FL",
  "Summer_Low_PA","Summer_Med_PA","Summer_High_PA"
)

abund <- matrix(rgamma(n_mags*length(samples2), shape=2, scale=50),
                nrow=n_mags, dimnames=list(mag_ids, samples2))

# Add structure: first 10 enriched in FL, next 10 enriched in PA
abund[1:10, 1:6] <- abund[1:10, 1:6] * 2
abund[11:20, 7:12] <- abund[11:20, 7:12] * 2

abundance <- as.data.frame(abund)
abundance_log <- log10(abundance + 1)

pheatmap(as.matrix(abundance_log),
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         border_color = "grey70",
         show_rownames = FALSE,
         main = "Clustered MAG Abundance Heatmap (Toy Data)")

# Save heatmap (base device)
#pdf("mag_heatmap_clustered_toy.pdf", width=10, height=9)
#pheatmap(as.matrix(abundance_log),
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         border_color = "grey70",
         show_rownames = FALSE,
         main = "Clustered MAG Abundance Heatmap (Toy Data)")
dev.off()

File: day9-visualization/scripts/python/heatmap.py
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

# ============================
# Create TOY MAG abundance data
# Rows = MAGs
# Columns = Samples
# Values = Abundance
# ============================

np.random.seed(42)

n_mags = 30
n_samples = 12

mags = [f"MAG_{i+1}" for i in range(n_mags)]
samples = [
    "Spring_Low_FL", "Spring_Med_FL", "Spring_High_FL",
    "Spring_Low_PA", "Spring_Med_PA", "Spring_High_PA",
    "Summer_Low_FL", "Summer_Med_FL", "Summer_High_FL",
    "Summer_Low_PA", "Summer_Med_PA", "Summer_High_PA"
]

# Generate abundance with structure
data = np.random.gamma(shape=2, scale=50, size=(n_mags, n_samples))

# Introduce ecological structure:
# First 10 MAGs enriched in FL
data[:10, :6] *= 2

# Next 10 MAGs enriched in PA
data[10:20, 6:] *= 2

abundance = pd.DataFrame(data, index=mags, columns=samples)

print("Toy MAG abundance preview:")
print(abundance.head())


# ============================
# Log transform
# ============================

abundance_log = np.log10(abundance + 1)


# ============================
# Clustered heatmap
# ============================

sns.clustermap(
    abundance_log,
    cmap='viridis',
    figsize=(12, 10),
    cbar_kws={'label': 'log10(Abundance + 1)'},
    linewidths=0.3,
    linecolor='gray',
    xticklabels=True,
    yticklabels=False,
    dendrogram_ratio=(0.1, 0.2),
    cbar_pos=(0.02, 0.8, 0.03, 0.15)
)

plt.savefig('mag_heatmap_clustered_toy.pdf', dpi=300, bbox_inches='tight')
plt.savefig('mag_heatmap_clustered_toy.png', dpi=300, bbox_inches='tight')

plt.show()

